File: day02.py
 
data  = [ int(x) for x in open("day02.dat").read().split(",")]

def run_program(program):
    ip = 0
    while True:
        op = program[ip]
        if op == 99:
            break
        a = program[ program[ip+1] ]
        b = program[ program[ip+2] ]
        dest = program[ip+3]
        if op == 1:
            program[dest] =  a + b
        else:
            program[dest] = a * b
        ip = ip + 4
    return program[0]

def part1(data):
    program = data[:]
    program[1]=12
    program[2]=2
    return run_program((program))


def part2(data):
    results = []
    for noun in range(100):
        for verb in range(100):
            program = data[:]
            program[1] = noun
            program[2] = verb
            r = run_program(program)
            results.append( (100*noun+verb, r))
    ans = [ x for x in results if x[1]==19690720]
    return ans
    
print(part1(data))
print(part2(data))

File: day02.clj
(ns day02
  (:require [clojure.string :as string]
            [utils :as u]))



;;(def raw_data (slurp "day02-sample.dat"))
(def raw_data (slurp "day02.dat"))
(def data  (->> (string/split raw_data #",")
                (map  u/parse-int)
                (into [])))

(defn do-op [program current]
  (let [op (get program current)
        a (get program  (get program (+ current 1)))
        b (get program  (get program (+ current 2)))
        c (get program (+ current 3))
        result (if (= op 1)
                 (+ a b)
                 (* a b))]
    (assoc program c result )
    ))

(defn process-part1 [program current]  
  (let [op (get program current)]
    (cond
      (= op 99) program
      :else (recur (do-op program current) (+ current 4))
      )
    ))

(defn process-part2 [program]
  (for [noun (range 100)
        verb (range 100)]
    (let [program (-> program
                      (assoc 1 noun)
                      (assoc 2 verb))]
      [(first (process-part1 program 0)) (+ (* 100 noun) verb)])
    ))

(defn part1[data]
  (let [program (-> data
                    (assoc 1 12)
                    (assoc 2 2))]
    (first  (process-part1 program 0))
    ))


(defn part2[data]
  (let [sols ( process-part2 data)
        ans (filter #(= (first %) 19690720) sols)
        ]
    ans
    )
  )
